File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/library/items/media/files/menu/FileListItemMenuBuilder.java
package com.lasthopesoftware.bluewater.client.library.items.media.files.menu;

import android.content.Context;
import android.content.Intent;
import android.graphics.Typeface;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.lasthopesoftware.bluewater.R;
import com.lasthopesoftware.bluewater.client.library.items.media.files.BaseMenuViewHolder;
import com.lasthopesoftware.bluewater.client.library.items.media.files.FilePlayClickListener;
import com.lasthopesoftware.bluewater.client.library.items.media.files.ServiceFile;
import com.lasthopesoftware.bluewater.client.library.items.media.files.ViewFileDetailsClickListener;
import com.lasthopesoftware.bluewater.client.library.items.media.files.nowplaying.INowPlayingFileProvider;
import com.lasthopesoftware.bluewater.client.library.items.menu.AbstractListItemMenuBuilder;
import com.lasthopesoftware.bluewater.client.library.items.menu.LongClickViewAnimatorListener;
import com.lasthopesoftware.bluewater.client.library.items.menu.NotifyOnFlipViewAnimator;
import com.lasthopesoftware.bluewater.client.library.items.menu.handlers.AbstractMenuClickHandler;
import com.lasthopesoftware.bluewater.client.playback.service.PlaybackService;
import com.lasthopesoftware.bluewater.client.playback.service.broadcasters.PlaylistEvents;
import com.lasthopesoftware.bluewater.shared.view.LazyViewFinder;
import com.lasthopesoftware.bluewater.shared.view.ViewUtils;
import com.lasthopesoftware.promises.Promise;
import com.vedsoft.futures.callables.VoidFunc;

import java.util.List;
import java.util.Map;

/**
 * Created by david on 11/7/15.
 */
public class FileListItemMenuBuilder extends AbstractListItemMenuBuilder<ServiceFile> {

    private final List<ServiceFile> serviceFiles;
    private final INowPlayingFileProvider nowPlayingFileProvider;

    private static final class ViewHolder extends BaseMenuViewHolder {
        final LazyViewFinder<ImageButton> addButtonFinder;

        ViewHolder(final FileListItemContainer fileListItemContainer, final LazyViewFinder<ImageButton> viewFileDetailsButtonFinder, final LazyViewFinder<ImageButton> playButtonFinder, final LazyViewFinder<ImageButton> addButtonFinder) {
            super(viewFileDetailsButtonFinder, playButtonFinder);

            this.addButtonFinder = addButtonFinder;
            this.fileListItemContainer = fileListItemContainer;
        }

        final FileListItemContainer fileListItemContainer;
        AbstractFileListItemNowPlayingHandler fileListItemNowPlayingHandler;
        public Promise<Map<String, String>> filePropertiesProvider;

        final ImageButton getAddButton() {
            return addButtonFinder.findView();
        }
    }

    public FileListItemMenuBuilder(final List<ServiceFile> serviceFiles, INowPlayingFileProvider nowPlayingFileProvider) {
        this.serviceFiles = serviceFiles;
        this.nowPlayingFileProvider = nowPlayingFileProvider;
    }

    @Override
    public View getView(int position, final ServiceFile serviceFile, View convertView, ViewGroup parent) {
        if (convertView == null) {
            final FileListItemContainer fileItemMenu = new FileListItemContainer(parent.getContext());
            final NotifyOnFlipViewAnimator notifyOnFlipViewAnimator = fileItemMenu.getViewAnimator();
            convertView = notifyOnFlipViewAnimator;

            final LayoutInflater inflater = (LayoutInflater) parent.getContext().getSystemService(Context.LAYOUT_INFLATER_SERVICE);

            final LinearLayout fileMenu = (LinearLayout)inflater.inflate(R.layout.layout_file_item_menu, parent, false);

            notifyOnFlipViewAnimator.addView(fileMenu);

            notifyOnFlipViewAnimator.setTag(
                    new ViewHolder(
                            fileItemMenu,
                            new LazyViewFinder<>(fileMenu, R.id.btnViewFileDetails),
                            new LazyViewFinder<>(fileMenu, R.id.btnPlaySong),
                            new LazyViewFinder<>(fileMenu, R.id.btnAddToPlaylist)));

            notifyOnFlipViewAnimator.setViewChangedListener(getOnViewChangedListener());
        }

        final ViewHolder viewHolder = (ViewHolder)convertView.getTag();

        final FileListItemContainer fileListItem = viewHolder.fileListItemContainer;

        final TextView textView = fileListItem.getTextViewFinder();

        if (viewHolder.filePropertiesProvider != null) viewHolder.filePropertiesProvider.cancel();
        viewHolder.filePropertiesProvider = FileNameTextViewSetter.startNew(serviceFile, textView);

        textView.setTypeface(null, Typeface.NORMAL);
        nowPlayingFileProvider
            .getNowPlayingFile()
            .then(VoidFunc.runCarelessly(f ->
                textView.setTypeface(null, ViewUtils.getActiveListItemTextViewStyle(serviceFile.getKey() == f.getKey()))));

        if (viewHolder.fileListItemNowPlayingHandler != null) viewHolder.fileListItemNowPlayingHandler.release();
        viewHolder.fileListItemNowPlayingHandler = new AbstractFileListItemNowPlayingHandler(fileListItem) {
            @Override
            public void onReceive(Context context, Intent intent) {
                final int fileKey = intent.getIntExtra(PlaylistEvents.PlaybackFileParameters.fileKey, -1);
                textView.setTypeface(null, ViewUtils.getActiveListItemTextViewStyle(serviceFile.getKey() == fileKey));
            }
        };

        final NotifyOnFlipViewAnimator viewAnimator = fileListItem.getViewAnimator();
        LongClickViewAnimatorListener.tryFlipToPreviousView(viewAnimator);
        viewHolder.getPlayButton().setOnClickListener(new FilePlayClickListener(viewAnimator, position, serviceFiles));
        viewHolder.getViewFileDetailsButton().setOnClickListener(new ViewFileDetailsClickListener(viewAnimator, serviceFile));
        viewHolder.getAddButton().setOnClickListener(new AddClickListener(viewAnimator, serviceFile));

        return viewAnimator;
    }

    private static class AddClickListener extends AbstractMenuClickHandler {
        private final ServiceFile mServiceFile;

        AddClickListener(NotifyOnFlipViewAnimator viewFlipper, ServiceFile serviceFile) {
            super(viewFlipper);
            mServiceFile = serviceFile;
        }

        @Override
        public void onClick(final View view) {
	        PlaybackService.addFileToPlaylist(view.getContext(), mServiceFile.getKey());

            super.onClick(view);
        }
    }
}

File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/browsing/items/media/files/menu/FileListItemMenuBuilder.kt
package com.lasthopesoftware.bluewater.client.browsing.items.media.files.menu

import android.content.Context
import android.content.Intent
import android.graphics.Typeface
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageButton
import android.widget.LinearLayout
import androidx.recyclerview.widget.RecyclerView
import com.lasthopesoftware.bluewater.R
import com.lasthopesoftware.bluewater.client.browsing.items.media.files.FilePlayClickListener
import com.lasthopesoftware.bluewater.client.browsing.items.media.files.ServiceFile
import com.lasthopesoftware.bluewater.client.browsing.items.media.files.details.ViewFileDetailsClickListener
import com.lasthopesoftware.bluewater.client.browsing.items.menu.AbstractViewChangedListenerContainer
import com.lasthopesoftware.bluewater.client.browsing.items.menu.LongClickViewAnimatorListener
import com.lasthopesoftware.bluewater.client.browsing.items.menu.NotifyOnFlipViewAnimator
import com.lasthopesoftware.bluewater.client.browsing.items.menu.handlers.AbstractMenuClickHandler
import com.lasthopesoftware.bluewater.client.playback.service.PlaybackService
import com.lasthopesoftware.bluewater.client.playback.service.broadcasters.PlaylistEvents
import com.lasthopesoftware.bluewater.client.playback.view.nowplaying.INowPlayingFileProvider
import com.lasthopesoftware.bluewater.shared.android.view.LazyViewFinder
import com.lasthopesoftware.bluewater.shared.android.view.ViewUtils
import com.lasthopesoftware.bluewater.shared.promises.extensions.LoopedInPromise
import com.namehillsoftware.handoff.promises.Promise

class FileListItemMenuBuilder(private val serviceFiles: List<ServiceFile>, private val nowPlayingFileProvider: INowPlayingFileProvider) : AbstractViewChangedListenerContainer() {

	fun getView(parent: ViewGroup): ViewHolder {
		val fileItemMenu = FileListItemContainer(parent.context)
		return ViewHolder(fileItemMenu)
	}

	inner class ViewHolder internal constructor(val fileListItemContainer: FileListItemContainer) : RecyclerView.ViewHolder(fileListItemContainer.viewAnimator) {
		private var fileListItemNowPlayingHandler: AbstractFileListItemNowPlayingHandler? = null
		private var promisedTextViewUpdate: Promise<*>? = null

		private val fileNameTextViewSetter = FileNameTextViewSetter(fileListItemContainer.findTextView())
		private val viewFileDetailsButtonFinder: LazyViewFinder<ImageButton>
		private val playButtonFinder: LazyViewFinder<ImageButton>
		private val addButtonFinder: LazyViewFinder<ImageButton>
		private val textView = fileListItemContainer.findTextView()

		init {
			val notifyOnFlipViewAnimator = fileListItemContainer.viewAnimator
			val inflater = notifyOnFlipViewAnimator.context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
			val fileMenu = inflater.inflate(R.layout.layout_file_item_menu, notifyOnFlipViewAnimator, false) as LinearLayout
			viewFileDetailsButtonFinder = LazyViewFinder(fileMenu, R.id.btnViewFileDetails)
			playButtonFinder = LazyViewFinder(fileMenu, R.id.btnPlaySong)
			addButtonFinder = LazyViewFinder(fileMenu, R.id.btnAddToPlaylist)

			notifyOnFlipViewAnimator.addView(fileMenu)
			notifyOnFlipViewAnimator.setViewChangedListener(getOnViewChangedListener())
		}

		fun update(position: Int) {
			val serviceFile = serviceFiles[position]

			promisedTextViewUpdate = fileNameTextViewSetter.promiseTextViewUpdate(serviceFile)

			textView.setTypeface(null, Typeface.NORMAL)
			nowPlayingFileProvider
				.nowPlayingFile
				.eventually<Unit>(LoopedInPromise.response({ f ->
					textView.setTypeface(null, ViewUtils.getActiveListItemTextViewStyle(serviceFile == f))
				}, textView.context))

			fileListItemNowPlayingHandler?.release()
			fileListItemNowPlayingHandler = object : AbstractFileListItemNowPlayingHandler(fileListItemContainer) {
				override fun onReceive(context: Context, intent: Intent) {
					val fileKey = intent.getIntExtra(PlaylistEvents.PlaybackFileParameters.fileKey, -1)
					textView.setTypeface(null, ViewUtils.getActiveListItemTextViewStyle(serviceFile.key == fileKey))
				}
			}

			val viewAnimator = fileListItemContainer.viewAnimator
			LongClickViewAnimatorListener.tryFlipToPreviousView(viewAnimator)
			playButtonFinder.findView().setOnClickListener(FilePlayClickListener(viewAnimator, position, serviceFiles))
			viewFileDetailsButtonFinder.findView().setOnClickListener(ViewFileDetailsClickListener(viewAnimator, serviceFile))
			addButtonFinder.findView().setOnClickListener(AddClickListener(viewAnimator, serviceFile))
		}
	}

	private class AddClickListener internal constructor(viewFlipper: NotifyOnFlipViewAnimator?, private val mServiceFile: ServiceFile) : AbstractMenuClickHandler(viewFlipper) {
		override fun onClick(view: View) {
			PlaybackService.addFileToPlaylist(view.context, mServiceFile.key)
			super.onClick(view)
		}
	}
}
